pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/asm/AsmSymbolResolver.java

```diff
@@ -105,8 +105,8 @@ URL getUrlOfInternalName(String internalName) {
             if (prev != failed && prev != null) {
                 return prev;
             }
-            @Nullable URL url = getUrlOfInternalName(iname);
-            Loader loader = url == null ? FailedLoader.INSTANCE : new UrlLoader(url);
+            @Nullable InputStream inputStream = getStreamOfInternalName(iname);
+            Loader loader = inputStream == null ? FailedLoader.INSTANCE : new StreamLoader(internalName, inputStream);
```

Comment:
I pushed changes in that direction. This is not a theoretical case, when running on pmd-core for instance with JDK 17, I get these stats:
```
Of 2250 queries to the classloader, 575 queries failed, 1159 classes were found and parsed successfully, 0 were found but failed parsing (!), 516 were found but never parsed.
```
So about a third of all classes found are never parsed.

Reply:
Now CloseResource shows a bunch of false positives related to TypeSystem and SymbolResolver

Reply:
Good point about closing these input streams.

Wouldn't it be easier to keep track of the InputStreams handed out by `ClasspathClassLoader#getResourceAsStream`? The ClasspathClassLoader will be closed at the end of the analysis (see `PmdAnalysis#close`). That way the change would be less invasive. And the javadoc of `URLClassLoader#getResourceAsStream` says:

> If this loader is closed, then any resources opened by this method will be closed.

Which means, URLClassLoader keeps track of any input stream it provided. We can do the same for the input streams, we provide via jrt-fs.

Reply:
Yes I think you're right this would be much easier! I will look into it

Reply:
I've done some tests, also using https://github.com/jenkinsci/lib-file-leak-detector and checking the sources: The InputStreams we return from JrtFileSystem are actually ByteArrayInputStreams (see https://github.com/openjdk/jdk/blob/970cd202049f592946f9c1004ea92dbd58abf6fb/src/java.base/share/classes/jdk/internal/jrtfs/JrtFileSystem.java#L334 ).

Since we now use URLClassLoader#getResourceAsStream to load the class resources, the URLClassLoader will eventually close the streams, even if we didn't consume them, together with the jar file, from which they are loaded.
(Before this PR, we used URL, which can leave the jar files open - see https://github.com/pmd/pmd/pull/4718#discussion_r1373146422)
So, we don't need to track the InputStreams ourselves, this is done already by URLClassLoader (see e.g. https://github.com/openjdk/jdk/blob/744e0893100d402b2b51762d57bcc2e99ab7fdcc/src/java.base/share/classes/java/net/URLClassLoader.java#L297-L308)

I manually verified, that we have no problems in this area now - we don't need to explicitly close our unloaded StreamLoaders, as long as we close our ClasspathClassLoader - which we do.

But I found two other file leaks (one with the JrtFileSystemProvider - which kept the jrt-fs.jar file open, another one in picocli).